File: sdks/python/test_sdk.py
"""
FastDataBroker Python SDK - Test Suite
Tests core functionality without requiring full server setup
"""

import time
import hashlib
import random
from dataclasses import dataclass
from typing import Dict, Any, Optional

# Mock SDK Classes
@dataclass(frozen=True)
class Message:
    topic: str
    payload: Any
    priority: int = 5
    ttl_seconds: int = 3600
    headers: Optional[Dict[str, str]] = None

@dataclass
class DeliveryResult:
    message_id: str
    status: str
    latency_ms: float
    timestamp: int

@dataclass
class ConnectionStats:
    is_connected: bool
    messages_sent: int
    messages_received: int
    connection_time_ms: int
    uptime_seconds: int
    last_message_time: int


class FastDataBrokerQuicClient:
    def __init__(self, config):
        self.config = config
        self.connected = False
        self.authenticated = False
        self.stats = {
            'messages_sent': 0,
            'messages_received': 0,
            'last_message_time': 0,
        }
        self.connection_start = 0
        self.message_handlers = {}

    def connect(self):
        print(f"Connecting to {self.config['host']}:{self.config['port']}...")
        self.connected = True
        self.connection_start = int(time.time() * 1000)
        print(f"✓ Connected to {self.config['host']}:{self.config['port']}")

    def send_message(self, message):
        if not self.connected:
            raise ConnectionError("Not connected")
        
        message_id = f"msg_{int(time.time() * 1000)}_{random.randint(0, 10000)}"
        latency = (time.time() % 50) + 5  # Simulate 5-55ms latency
        self.stats['messages_sent'] += 1
        self.stats['last_message_time'] = int(time.time() * 1000)
        
        return DeliveryResult(
            message_id=message_id,
            status='success',
            latency_ms=latency,
            timestamp=int(time.time() * 1000),
        )

    def on_message(self, topic, handler):
        self.message_handlers[topic] = handler

    def off_message(self, topic):
        if topic in self.message_handlers:
            del self.message_handlers[topic]

    def get_stats(self):
        uptime_ms = int(time.time() * 1000) - self.connection_start if self.connected else 0
        return ConnectionStats(
            is_connected=self.connected,
            messages_sent=self.stats['messages_sent'],
            messages_received=self.stats['messages_received'],
            connection_time_ms=uptime_ms,
            uptime_seconds=uptime_ms // 1000,
            last_message_time=self.stats['last_message_time'],
        )

    def is_connected(self):
        return self.connected

    def disconnect(self):
        self.connected = False
        print('✓ Disconnected')


# ============================================================================
# TEST SUITE
# ============================================================================

tests_passed = 0
tests_failed = 0


def run_test(name, test_fn):
    global tests_passed, tests_failed
    try:
        test_fn()
        print(f"✅ PASS: {name}")
        tests_passed += 1
    except Exception as error:
        print(f"❌ FAIL: {name}")
        print(f"   Error: {str(error)}")
        tests_failed += 1


def test1_basic_connection():
    config = {
        'host': 'localhost',
        'port': 6000,
        'tenant_id': 'test-tenant',
        'client_id': 'test-client',
        'psk_secret': 'test-secret',
    }

    client = FastDataBrokerQuicClient(config)

    if client.is_connected():
        raise AssertionError('Should not be connected yet')

    client.connect()

    if not client.is_connected():
        raise AssertionError('Should be connected')

    client.disconnect()


def test2_send_message():
    client = FastDataBrokerQuicClient({
        'host': 'localhost',
        'port': 6000,
        'tenant_id': 'test-tenant',
        'client_id': 'test-client',
        'psk_secret': 'test-secret',
    })

    client.connect()

    result = client.send_message(
        Message(topic='test.topic', payload={'data': 'test'}, priority=5)
    )

    if result.status != 'success':
        raise AssertionError(f"Expected status 'success', got '{result.status}'")

    if not result.message_id:
        raise AssertionError('Message ID should not be empty')

    if result.latency_ms < 0:
        raise AssertionError('Latency should be non-negative')

    client.disconnect()


def test3_message_handlers():
    client = FastDataBrokerQuicClient({
        'host': 'localhost',
        'port': 6000,
        'tenant_id': 'test-tenant',
        'client_id': 'test-client',
        'psk_secret': 'test-secret',
    })

    client.connect()

    handler_called = False
    
    def handler(message):
        nonlocal handler_called
        handler_called = True

    client.on_message('test.topic', handler)

    if 'test.topic' not in client.message_handlers:
        raise AssertionError('Handler should be registered')

    client.off_message('test.topic')

    if 'test.topic' in client.message_handlers:
        raise AssertionError('Handler should be unregistered')

    client.disconnect()


def test4_connection_statistics():
    client = FastDataBrokerQuicClient({
        'host': 'localhost',
        'port': 6000,
        'tenant_id': 'test-tenant',
        'client_id': 'test-client',
        'psk_secret': 'test-secret',
    })

    client.connect()

    # Send some messages
    for i in range(5):
        client.send_message(Message(topic='test.topic', payload={'index': i}))

    # Add small delay to ensure timing is measurable
    import time
    time.sleep(0.05)

    stats = client.get_stats()

    if not stats.is_connected:
        raise AssertionError('Should be connected')

    if stats.messages_sent != 5:
        raise AssertionError(f"Expected 5 messages sent, got {stats.messages_sent}")

    if stats.uptime_seconds < 0:
        raise AssertionError('Uptime should be non-negative')

    client.disconnect()


def test5_concurrent_messages():
    client = FastDataBrokerQuicClient({
        'host': 'localhost',
        'port': 6000,
        'tenant_id': 'test-tenant',
        'client_id': 'test-client',
        'psk_secret': 'test-secret',
    })

    client.connect()

    results = []
    for i in range(10):
        result = client.send_message(
            Message(topic='test.concurrent', payload={'index': i})
        )
        results.append(result)

    if len(results) != 10:
        raise AssertionError(f"Expected 10 results, got {len(results)}")

    if not all(r.status == 'success' for r in results):
        raise AssertionError('All messages should be successful')

    if client.stats['messages_sent'] != 10:
        raise AssertionError(f"Expected 10 messages sent, got {client.stats['messages_sent']}")

    client.disconnect()


def test6_priority_levels():
    client = FastDataBrokerQuicClient({
        'host': 'localhost',
        'port': 6000,
        'tenant_id': 'test-tenant',
        'client_id': 'test-client',
        'psk_secret': 'test-secret',
    })

    client.connect()

    priorities = [1, 5, 10, 20]  # LOW, NORMAL, HIGH, CRITICAL

    for priority in priorities:
        result = client.send_message(
            Message(
                topic='test.priority',
                payload={'priority': priority},
                priority=priority,
            )
        )

        if result.status != 'success':
            raise AssertionError(f"Failed to send message with priority {priority}")

    client.disconnect()


def test7_latency_measurement():
    client = FastDataBrokerQuicClient({
        'host': 'localhost',
        'port': 6000,
        'tenant_id': 'test-tenant',
        'client_id': 'test-client',
        'psk_secret': 'test-secret',
    })

    client.connect()

    latencies = []
    for i in range(50):
        result = client.send_message(
            Message(topic='test.latency', payload={'iteration': i})
        )
        latencies.append(result.latency_ms)

    avg_latency = sum(latencies) / len(latencies)
    max_latency = max(latencies)

    if avg_latency < 0:
        raise AssertionError('Average latency should be non-negative')

    print(f"   Average latency: {avg_latency:.2f}ms, Max: {max_latency:.2f}ms")

    client.disconnect()


def test8_error_handling():
    client = FastDataBrokerQuicClient({
        'host': 'localhost',
        'port': 6000,
        'tenant_id': 'test-tenant',
        'client_id': 'test-client',
        'psk_secret': 'test-secret',
    })

    try:
        client.send_message(Message(topic='test', payload={}))
        raise AssertionError('Should have thrown error')
    except ConnectionError as e:
        if str(e) != 'Not connected':
            raise AssertionError('Should throw "Not connected" error')

    client.connect()
    result = client.send_message(Message(topic='test', payload={}))
    if result.status != 'success':
        raise AssertionError('Should send successfully after connect')

    client.disconnect()


def test9_configuration_validation():
    config = {
        'host': 'localhost',
        'port': 6000,
        'tenant_id': 'test-tenant',
        'client_id': 'test-client',
        'psk_secret': 'test-secret',
    }

    client = FastDataBrokerQuicClient(config)

    if client.config['host'] != 'localhost':
        raise AssertionError('Configuration not properly saved')

    if client.config['port'] != 6000:
        raise AssertionError('Port configuration not properly saved')


# ============================================================================
# RUN ALL TESTS
# ============================================================================

def run_all_tests():
    global tests_passed, tests_failed
    
    print('\n' + '='*70)
    print('FastDataBroker Python SDK - Test Suite')
    print('='*70 + '\n')

    run_test('1. Basic Connection', test1_basic_connection)
    run_test('2. Send Message', test2_send_message)
    run_test('3. Message Handlers', test3_message_handlers)
    run_test('4. Connection Statistics', test4_connection_statistics)
    run_test('5. Concurrent Messages', test5_concurrent_messages)
    run_test('6. Priority Levels', test6_priority_levels)
    run_test('7. Latency Measurement', test7_latency_measurement)
    run_test('8. Error Handling', test8_error_handling)
    run_test('9. Configuration Validation', test9_configuration_validation)

    print('\n' + '='*70)
    print(f'Results: {tests_passed} passed, {tests_failed} failed')
    print('='*70 + '\n')

    return tests_failed == 0


if __name__ == '__main__':
    import sys
    success = run_all_tests()
    sys.exit(0 if success else 1)

File: sdks/java/TestSDK.java
package com.fastdatabroker.sdk.test;

import java.util.*;
import java.util.concurrent.*;

/**
 * FastDataBroker Java SDK - Test Suite
 */

// Mock SDK Types
class Message {
    public String topic;
    public Object payload;
    public int priority;
    public int ttlSeconds;
    public Map<String, String> headers;

    public Message(String topic, Object payload) {
        this.topic = topic;
        this.payload = payload;
        this.priority = 5;
        this.ttlSeconds = 3600;
        this.headers = new HashMap<>();
    }

    public Message withPriority(int priority) {
        this.priority = priority;
        return this;
    }
}

class DeliveryResult {
    public String messageId;
    public String status;
    public double latencyMs;
    public long timestamp;

    public DeliveryResult(String messageId, String status, double latencyMs, long timestamp) {
        this.messageId = messageId;
        this.status = status;
        this.latencyMs = latencyMs;
        this.timestamp = timestamp;
    }

    @Override
    public String toString() {
        return String.format("DeliveryResult{messageId='%s', status='%s', latencyMs=%.2f}", 
            messageId, status, latencyMs);
    }
}

class ConnectionStats {
    public boolean isConnected;
    public long messagesSent;
    public long messagesReceived;
    public long connectionTimeMs;
    public long uptimeSeconds;
    public long lastMessageTime;

    public ConnectionStats(boolean isConnected, long messagesSent, long messagesReceived,
                          long connectionTimeMs, long uptimeSeconds, long lastMessageTime) {
        this.isConnected = isConnected;
        this.messagesSent = messagesSent;
        this.messagesReceived = messagesReceived;
        this.connectionTimeMs = connectionTimeMs;
        this.uptimeSeconds = uptimeSeconds;
        this.lastMessageTime = lastMessageTime;
    }

    @Override
    public String toString() {
        return String.format("ConnectionStats{isConnected=%b, messagesSent=%d, uptimeSeconds=%d}", 
            isConnected, messagesSent, uptimeSeconds);
    }
}

class FastDataBrokerQuicClient {
    private Map<String, String> config;
    private boolean connected;
    private boolean authenticated;
    private Map<String, String> stats;
    private long connectionStart;
    private Map<String, Consumer<Object>> messageHandlers;

    @FunctionalInterface
    interface Consumer<T> {
        void accept(T t);
    }

    public FastDataBrokerQuicClient(String host, int port, String tenantId, String clientId, String pskSecret) {
        config = new HashMap<>();
        config.put("host", host);
        config.put("port", String.valueOf(port));
        config.put("tenant_id", tenantId);
        config.put("client_id", clientId);
        config.put("psk_secret", pskSecret);

        this.connected = false;
        this.authenticated = false;
        this.stats = new HashMap<>();
        stats.put("messages_sent", "0");
        stats.put("messages_received", "0");
        this.messageHandlers = new HashMap<>();
    }

    public void connect() throws Exception {
        String host = config.get("host");
        String port = config.get("port");
        System.out.printf("Connecting to %s:%s...%n", host, port);
        this.connected = true;
        this.connectionStart = System.currentTimeMillis();
        System.out.printf("✓ Connected to %s:%s%n", host, port);
    }

    public DeliveryResult sendMessage(Message message) throws Exception {
        if (!connected) {
            throw new Exception("Not connected");
        }

        String messageId = String.format("msg_%d_%d", System.currentTimeMillis(), 
            (int)(Math.random() * 10000));
        double latency = Math.random() * 50 + 5;
        long messagesSent = Long.parseLong(stats.getOrDefault("messages_sent", "0"));
        stats.put("messages_sent", String.valueOf(messagesSent + 1));

        return new DeliveryResult(messageId, "success", latency, System.currentTimeMillis());
    }

    public void onMessage(String topic, Consumer<Object> handler) {
        messageHandlers.put(topic, handler);
    }

    public void offMessage(String topic) {
        messageHandlers.remove(topic);
    }

    public ConnectionStats getStats() {
        long uptime = connected ? System.currentTimeMillis() - connectionStart : 0;
        return new ConnectionStats(
            connected,
            Long.parseLong(stats.getOrDefault("messages_sent", "0")),
            Long.parseLong(stats.getOrDefault("messages_received", "0")),
            uptime,
            uptime / 1000,
            0
        );
    }

    public boolean isConnected() {
        return connected;
    }

    public void disconnect() {
        connected = false;
        System.out.println("✓ Disconnected");
    }
}

/**
 * Test Suite
 */
public class TestSDK {
    private static int testsPassed = 0;
    private static int testsFailed = 0;

    static void runTest(String name, Runnable testFn) {
        try {
            testFn.run();
            System.out.printf("✅ PASS: %s%n", name);
            testsPassed++;
        } catch (Exception error) {
            System.out.printf("❌ FAIL: %s%n", name);
            System.out.printf("   Error: %s%n", error.getMessage());
            testsFailed++;
        }
    }

    static void test1BasicConnection() throws Exception {
        FastDataBrokerQuicClient client = new FastDataBrokerQuicClient(
            "localhost", 6000, "test-tenant", "test-client", "test-secret");

        if (client.isConnected()) {
            throw new Exception("Should not be connected yet");
        }

        client.connect();

        if (!client.isConnected()) {
            throw new Exception("Should be connected");
        }

        client.disconnect();
    }

    static void test2SendMessage() throws Exception {
        FastDataBrokerQuicClient client = new FastDataBrokerQuicClient(
            "localhost", 6000, "test-tenant", "test-client", "test-secret");

        client.connect();

        Message msg = new Message("test.topic", new HashMap<>()).withPriority(5);
        DeliveryResult result = client.sendMessage(msg);

        if (!result.status.equals("success")) {
            throw new Exception("Expected status 'success', got '" + result.status + "'");
        }

        if (result.messageId == null || result.messageId.isEmpty()) {
            throw new Exception("Message ID should not be empty");
        }

        if (result.latencyMs < 0) {
            throw new Exception("Latency should be non-negative");
        }

        client.disconnect();
    }

    static void test3MessageHandlers() throws Exception {
        FastDataBrokerQuicClient client = new FastDataBrokerQuicClient(
            "localhost", 6000, "test-tenant", "test-client", "test-secret");

        client.connect();

        client.onMessage("test.topic", msg -> {});

        if (!client.messageHandlers.containsKey("test.topic")) {
            throw new Exception("Handler should be registered");
        }

        client.offMessage("test.topic");

        if (client.messageHandlers.containsKey("test.topic")) {
            throw new Exception("Handler should be unregistered");
        }

        client.disconnect();
    }

    static void test4ConnectionStatistics() throws Exception {
        FastDataBrokerQuicClient client = new FastDataBrokerQuicClient(
            "localhost", 6000, "test-tenant", "test-client", "test-secret");

        client.connect();

        for (int i = 0; i < 5; i++) {
            client.sendMessage(new Message("test.topic", new HashMap<>()));
        }

        // Add small delay to ensure timing is measurable
        Thread.sleep(50);

        ConnectionStats stats = client.getStats();

        if (!stats.isConnected) {
            throw new Exception("Should be connected");
        }

        if (stats.messagesSent != 5) {
            throw new Exception("Expected 5 messages sent, got " + stats.messagesSent);
        }

        if (stats.uptimeSeconds < 0) {
            throw new Exception("Uptime should be non-negative");
        }

        client.disconnect();
    }

    static void test5ConcurrentMessages() throws Exception {
        FastDataBrokerQuicClient client = new FastDataBrokerQuicClient(
            "localhost", 6000, "test-tenant", "test-client", "test-secret");

        client.connect();

        List<DeliveryResult> results = new ArrayList<>();
        for (int i = 0; i < 10; i++) {
            results.add(client.sendMessage(new Message("test.concurrent", new HashMap<>())));
        }

        if (results.size() != 10) {
            throw new Exception("Expected 10 results, got " + results.size());
        }

        for (DeliveryResult result : results) {
            if (!result.status.equals("success")) {
                throw new Exception("All messages should be successful");
            }
        }

        if (client.stats.get("messages_sent") == null || 
            !client.stats.get("messages_sent").equals("10")) {
            throw new Exception("Expected 10 messages sent");
        }

        client.disconnect();
    }

    static void test6PriorityLevels() throws Exception {
        FastDataBrokerQuicClient client = new FastDataBrokerQuicClient(
            "localhost", 6000, "test-tenant", "test-client", "test-secret");

        client.connect();

        int[] priorities = {1, 5, 10, 20};

        for (int priority : priorities) {
            DeliveryResult result = client.sendMessage(
                new Message("test.priority", new HashMap<>()).withPriority(priority)
            );

            if (!result.status.equals("success")) {
                throw new Exception("Failed to send message with priority " + priority);
            }
        }

        client.disconnect();
    }

    static void test7LatencyMeasurement() throws Exception {
        FastDataBrokerQuicClient client = new FastDataBrokerQuicClient(
            "localhost", 6000, "test-tenant", "test-client", "test-secret");

        client.connect();

        List<Double> latencies = new ArrayList<>();
        for (int i = 0; i < 50; i++) {
            DeliveryResult result = client.sendMessage(
                new Message("test.latency", new HashMap<>())
            );
            latencies.add(result.latencyMs);
        }

        double avgLatency = 0;
        for (double latency : latencies) {
            avgLatency += latency;
        }
        avgLatency /= latencies.size();

        if (avgLatency < 0) {
            throw new Exception("Average latency should be non-negative");
        }

        double maxLatency = Collections.max(latencies);

        System.out.printf("   Average latency: %.2fms, Max: %.2fms%n", avgLatency, maxLatency);

        client.disconnect();
    }

    static void test8ErrorHandling() throws Exception {
        FastDataBrokerQuicClient client = new FastDataBrokerQuicClient(
            "localhost", 6000, "test-tenant", "test-client", "test-secret");

        try {
            client.sendMessage(new Message("test", new HashMap<>()));
            throw new Exception("Should have thrown error");
        } catch (Exception e) {
            if (!e.getMessage().equals("Not connected")) {
                throw new Exception("Should throw 'Not connected' error");
            }
        }

        client.connect();
        DeliveryResult result = client.sendMessage(new Message("test", new HashMap<>()));
        if (!result.status.equals("success")) {
            throw new Exception("Should send successfully after connect");
        }

        client.disconnect();
    }

    static void test9ConfigurationValidation() throws Exception {
        FastDataBrokerQuicClient client = new FastDataBrokerQuicClient(
            "localhost", 6000, "test-tenant", "test-client", "test-secret");

        if (!client.config.get("host").equals("localhost")) {
            throw new Exception("Host configuration not properly saved");
        }

        if (!client.config.get("port").equals("6000")) {
            throw new Exception("Port configuration not properly saved");
        }
    }

    public static void main(String[] args) {
        System.out.println("\n" + "=".repeat(70));
        System.out.println("FastDataBroker Java SDK - Test Suite");
        System.out.println("=".repeat(70) + "\n");

        runTest("1. Basic Connection", () -> test1BasicConnection());
        runTest("2. Send Message", () -> test2SendMessage());
        runTest("3. Message Handlers", () -> test3MessageHandlers());
        runTest("4. Connection Statistics", () -> test4ConnectionStatistics());
        runTest("5. Concurrent Messages", () -> test5ConcurrentMessages());
        runTest("6. Priority Levels", () -> test6PriorityLevels());
        runTest("7. Latency Measurement", () -> test7LatencyMeasurement());
        runTest("8. Error Handling", () -> test8ErrorHandling());
        runTest("9. Configuration Validation", () -> test9ConfigurationValidation());

        System.out.println("\n" + "=".repeat(70));
        System.out.printf("Results: %d passed, %d failed%n", testsPassed, testsFailed);
        System.out.println("=".repeat(70) + "\n");

        System.exit(testsFailed > 0 ? 1 : 0);
    }
}
